docs/lsp.md

```diff
@@ -0,0 +1,480 @@
+
+#### Example sequence diagram
+
+```mermaid
+
+sequenceDiagram
+
+participant Client
+participant FsLSP
+participant WorkspaceStateManagement
+participant FCS
+
+Client->>FsLSP: textDocument/didChange
+FsLSP->>WorkspaceStateManagement: Update workspace state
+Client->>+FsLSP: textDocument/diagnostic
+FsLSP->>+WorkspaceStateManagement: Get current workspace snapshot
+WorkspaceStateManagement-->>-FsLSP: Current workspace snapshot
+FsLSP->>+FCS: ParseAndCheckFileInProject
+FCS-->>-FsLSP: FSharpCheckFileResults
+FsLSP-->>-Client: DocumentDiagnosticReport
+
+```
+
+
+#### Concurrency and cancellation
+
+We need to make sure when a project snapshot is needed to process a request, it will always contain all change notifications that were received before. There shouldn't be need for a global request queue, possibly only for change notifications inside WorkspaceStateManagement. And any language requests will wait until WorkspaceStateManagement has absorbed all previous changes.
+
+For cancellation we should decide, potentially on a case-by-case basis what to do when an incoming change notification invalidates an ongoing request.
+
+1. Do nothing and let the request run either until it completes or is cancelled by the client. This is the easiest so we will probably start with this.
+2. Cancel the request and return "cancelled" result to the client. The client then needs to deal with it.
+3. Internally restart the work updated inputs and complete the original request until cancelled. This mode might save a bunch of back-and-forth with the client when user is typing. But it requires the client to expect it and not send cancellations a re-requests for the same thing.
+
+<details>
+<summary>Request update example</summary>
+
+```mermaid
+
+sequenceDiagram
+
+participant Client
+participant FsLSP
+participant WorkspaceStateManagement
+participant FCS
+
+Client->>+FsLSP: textDocument/diagnostic
+FsLSP->>+WorkspaceStateManagement: Get current workspace snapshot
+WorkspaceStateManagement-->>-FsLSP: Current workspace snapshot
+FsLSP->>FCS: ParseAndCheckFileInProject
+activate FCS
+Client->>FsLSP: textDocument/didChange
+FsLSP->>FCS: Cancel ParseAndCheckFileInProject
+deactivate FCS
+FsLSP->>WorkspaceStateManagement: Update workspace state
+FsLSP->>+WorkspaceStateManagement: Get current workspace snapshot
+WorkspaceStateManagement-->>-FsLSP: Current workspace snapshot
+
+FsLSP->>+FCS: ParseAndCheckFileInProject
+FCS-->>-FsLSP: FSharpCheckFileResults
+FsLSP-->>-Client: DocumentDiagnosticReport
+
+```
+
+</details>
+
+
+#### Request tracking
+
+This will hopefully be handled by the LSP library. Although we might need to be able to keep track of requests in progress for the purpose of cancellation.
+
+### NewVsix
+
+New VS extension that will be backed by the LSP server. It should mostly contain plumbing to connect LSP to the editor and CPS.
+
+> Is there a library available to simplify some of this?
```

Comment:
maybe https://github.com/VsixCommunity/Community.VisualStudio.Toolkit

Reply:
It is very unlikely we'll be able to take such big 3rd party dependency. Especially one which is VS-related, since we will be blocked each time VS updates extensibility layer (e.f. updates to dev18 or similar).